[Enhancement]: Add interface `location.broadcast_domain` parameter

acch opened the pull request:
Adds ability to configure `location.broadcast_domain` parameter of IP Interfaces.

Closes #305.

Acceptance tests pass:
```shell
$ TF_ACC=1 go test ./internal/provider/networking/network_ip_interface_resource_test.go -v
=== RUN   TestAccNetworkIpInterfaceResource
--- PASS: TestAccNetworkIpInterfaceResource (10.65s)
PASS
ok      command-line-arguments  10.664s
```
```shell
$ TF_ACC=1 go test ./internal/provider/networking/network_ip_interface_resource_alias_test.go -v
=== RUN   TestAccNetworkIpInterfaceResourceAlias
--- PASS: TestAccNetworkIpInterfaceResourceAlias (9.21s)
PASS
ok      command-line-arguments  9.225s
```

Example Terraform Configurations:
```terraform
resource "netapp-ontap_network_ip_interface" "with_node_port" {
  cx_profile_name = ...
  name            = ...
  svm_name        = ...

  ip = { ... }

  location = {
    home_port = "e0a-100"
    home_node = "node1"
  }
}
```

```terraform
resource "netapp-ontap_network_ip_interface" "with_broadcast_domain" {
  cx_profile_name = ...
  name            = ...
  svm_name        = ...

  ip = { ... }

  location = {
    broadcast_domain = {
      name = "bd1"
    }
  }
}
```

acch committed:
feat: add `location.broadcast_domain` to IP interface data source

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add `location.broadcast_domain` to IP interfaces data source

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: make existing location attributes optional

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add `location.broadcast_domain` to IP interfaces resource

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add examples for cluster-, and svm-scoped IP interfaces

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add examples for using node & port-, and broadcast domain for location

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
fix: add home port & node to POST request when defined, even if unchanged

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add acceptance tests for 'location' attribute combinations

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
chore: change attributes for consistency with other examples

Signed-off-by: Achim Christ <achim.christ@sva.de>

carchi8py commented on internal/provider/networking/network_ip_interface_resource.go:
```diff
@@ -107,16 +122,40 @@ func (r *IPInterfaceResource) Schema(ctx context.Context, req resource.SchemaReq
+									stringplanmodifier.RequiresReplace(),
+								},
+							},
+							"id": schema.StringAttribute{
```
We typically don't have UUID in resources other than the top level one. As we only expect users to know the names and not the internal ID.

We do the top level one as computed as we do need a unique value for terraform. 

So this ID can probably be removed for the Resouce. (For the data source it can stay)

acch replied:
The ID is optional, so users don't need to use it. However, referencing a broadcast domain by ID has the following advantage: the ID is computed by the `broadcast_domain` resource, it is "unknown" during plan-phase and only has a known value after the broadcast domain exists. Hence, Terraform implicitly understands that the `broadcast_domain` resource needs to be created first, and the interface can only be created *after* the `broadcast_domain`. When referencing a `broadcast_domain` by name, users would need to explicitly add this dependency with the `depends_on` meta-argument.

Hence, I'd suggest to keep the UUID as an optional attribute - do you agree?

carchi8py reviewed (changes requested):
Looks good as well just missing the documentation updates.

acch committed:
chore: add `location.broadcast_domain` parameter to attribute list

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
Thanks @carchi8py for your feedback!

I've added the `location.broadcast_domain` argument to the docs. Please let me know what else is missing.

acch commented:
@carchi8py Are we good to merge this PR?

Please let me know what I need to do exactly to have the PRs added to the provider. My understanding is that I've addressed all requested changes... please let me know what's missing.

Many thanks!

suhasbshekar commented:
Can you update CHANGELOG.md file?

Also your branch is behind the integration/main, we added few more checks to the pipeline as it was missing here, since your PR is from a fork of this repo, so could you pull the new changes into your branch and update the PR?

Just for our reference, since we have disabled ACC tests as our testing profiles are down, acc tests are passing for this PR when tested locally.

acch committed:
chore: update changelog

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
@suhasbshekar Thanks for your feedback.

- I've added an entry to CHANGELOG.md
- I've rebased my branch to integration/main (once again)

Acc tests are passing on my local test system (ONTAP Select running 9.15.1P4):

```shell
$ date && TF_ACC=1 go test ./internal/provider/networking/network_ip_interface_resource_test.go -v && date
Tue Feb  4 10:03:55 UTC 2025
=== RUN   TestAccNetworkIpInterfaceResource
--- PASS: TestAccNetworkIpInterfaceResource (7.16s)
PASS
ok      command-line-arguments  7.164s
Tue Feb  4 10:04:03 UTC 2025
```

Please let me know what else is needed.
Many thanks!

suhasbshekar commented:
Please resolve the conflicts.

acch commented:
@suhasbshekar I've rebased the PR once again.

Can you please let me know when you'll be able to merge these changes?

acch committed:
fix: correct syntax

acch committed:
fix: merge test cases
